include/pybind11/embed.h

```diff
@@ -83,29 +85,101 @@ struct embedded_module {
     }
 };
 
+struct wide_char_arg_deleter {
+    void operator()(wchar_t* ptr) const {
+#if PY_VERSION_HEX >= 0x030500f0
+        // API docs: https://docs.python.org/3/c-api/sys.html#c.Py_DecodeLocale
+        PyMem_RawFree(ptr);
+#else
+        delete[] ptr;
+#endif
+    }
+};
+
+inline wchar_t* widen_chars(const char* safe_arg) {
+#if PY_VERSION_HEX >= 0x030500f0
+    wchar_t* widened_arg = Py_DecodeLocale(safe_arg, nullptr);
+#else
+    wchar_t* widened_arg = nullptr;
+#  if defined(HAVE_BROKEN_MBSTOWCS) && HAVE_BROKEN_MBSTOWCS
+    size_t count = strlen(safe_arg);
+#  else
+    size_t count = mbstowcs(nullptr, safe_arg, 0);
+#  endif
+    if (count != static_cast<size_t>(-1)) {
+        widened_arg = new wchar_t[count + 1];
+        mbstowcs(widened_arg, safe_arg, count + 1);
+    }
+#endif
+    return widened_arg;
+}
+
+/// Python 2.x/3.x-compatible version of `PySys_SetArgv`
+inline void set_interpreter_argv(int argc, const char* const* argv, bool add_program_dir_to_path) {
+    // Before it was special-cased in python 3.8, passing an empty or null argv
+    // caused a segfault, so we have to reimplement the special case ourselves.
+    bool special_case = (argv == nullptr || argc <= 0);
+
+    const char* const empty_argv[] {"\0"};
+    const char* const* safe_argv = special_case ? empty_argv : argv;
+    if (special_case)
+        argc = 1;
+
+#if PY_MAJOR_VERSION >= 3
+    auto argv_size = static_cast<size_t>(argc);
+    // SetArgv* on python 3 takes wchar_t, so we have to convert.
+    std::unique_ptr<wchar_t*[]> widened_argv(new wchar_t*[argv_size]);
+    std::vector< std::unique_ptr<wchar_t[], wide_char_arg_deleter> > widened_argv_entries;
```

Comment:
Any particular reason this is an std::vector instead of c or std::array?

Reply:
and if you are going to use it, you should call reserve() to reserve the space.

Reply:
We can't use `std::array` because `argv_size` is not known at compile time. Using a C type would just require more manual memory management. Your commit adding `.reserve()` to the vector is the correct solution.